### DISCUSSION TITLE
Netmiko find prompt before applying configuration

### DISCUSSION THREAD

**Discussion by kbSean1** [ktbyers/netmiko#2325] [General]
I use Netmiko to ssh to a cisco router and  I was trying to apply the BGP configuration using ssh_conn.send_config_set(). Below is a sample configuration to be applied. Expected behavior:
bgp router-id x.x.x.x should be configured under the vrf address family. For some reason, when I tested manually, router didn't change the prompt and hence applied all the configuration to the global mode instead of address family mode and resulted in a BGP flap. Is there any way to check the prompt before applying the configuration?
router bgp as_no
  address-family ipv4 vrf {vrf_name}
      bgp router-id x.x.x.x

**Comment by ktbyers** [comment]
@kbSean1 Netmiko will send what you specified via send_config_set().

Do you know why it didn't go into the address-family VRF? Was it because the VRF didn't exist?

You can use the `error_pattern` feature in send_config_set to cause configuration commands to abort if specific error messages are detected. So you could abort if you failed to enter into the VRF (for example, for a case like the following):

```
cisco3(config-router)#address-family ipv4 vrf foo 
% VRF foo does not exist.
cisco3(config-router)#
```

I don't have a document on this, but here is a video I made on this not to long ago:

https://youtu.be/oGtCebtgUio

**Reply by kbSean1** [reply]
@ktbyers VRF did exist. There are prechecks to ensure if the VRF is already present or not before applying BGP configuration command.  To be more specific, this is what happened.
R1(config)#router bgp asn
R1(config-router)#
R1(config-router)#no address-family ipv4 vrf vrf_name
R1(config-router)# address-family ipv4 vrf vrf_name
% Topology f959c5b8b::VPNv4 Unicast::base is currently being deconfigured.
R1(config-router)#  bgp router-id x.x.x.x
If this is the case, would error_pattern feature work or is there any other alternative to change the configuration from one mode to another , get the current prompt from the device and then apply the commands?

**Reply by ktbyers** [reply]
You could set `error_pattern` to r"% Topology.*".

If that error message was detected, then a `ConfigInvalidException` would be raised and none of the other configuration commands would be sent.

So in your case it would end with this command:

```
R1(config-router)# address-family ipv4 vrf vrf_name
% Topology f959c5b8b::VPNv4 Unicast::base is currently being deconfigured.
```

Your code would then need to handle what to do in the failure scenario (if anything).

This feature was introduced in Netmiko 3.4.0

**Reply by kbSean1** [reply]
@ktbyers Thanks a lot for helping with this issue. This works perfectly. I was also wondering if there is any other alternative to check the prompt before applying configuration.

**Reply by kbSean1** [reply]
For instance in this example, 
When we apply :
R1(config-router)# address-family ipv4 vrf vrf_name 
this would change to 
R1(config-router-af)#
Is there a way we can find this prompt before applying next commands?

**Reply by ktbyers** [reply]
Sure you can write a custom function using write_channel and read_channel and pattern match based on that.

I also just put in the `develop` branch a `send_command_multiline` which allows you to send a series of commands and corresponding expected patterns:

```python
    commands = ( 
        ("ping", r"ip"),
        ("", r"Target IP address"),
        ("8.8.8.8", "Repeat count"),
        ("100", "Datagram size"),
        ("", "Timeout in seconds"),
        ("", "Extended"),
        ("", "Sweep"),
        ("", ""),
    )   
    output = net_connect.send_multiline(commands)
```

The first element of the tuple is a command to send. The second element is the pattern to expect in the output.

If the command is a null-string, then that is equivalent to sending an <enter>. If the pattern is a null-string, then it will base the pattern on the prompt.

**Reply by kbSean1** [reply]
@ktbyers Thanks! When I did some more testing today in the Cisco router, I saw this kind of behavior:

Manual execution results:
R1#conf t
Enter configuration commands, one per line. End with CNTL/Z.
R1(config)#router bgp {as_no}
R1(config-router)#address-family ipv4 vrf VRF_WL
% VRF VRF_WL does not exist.
R1(config-router)#exit
Assuming initial cluster id to be 0.0.16.0,
python code:
config_lines_1 = ["router bgp {as_no}","address-family ipv4 vrf VRF_WL","bgp cluster-id  0.0.17.2"]
output1 = ssh_conn.send_config_set(config_lines_1, error_pattern= r"%.*") 
After execution, resulted in the below error.
  
   line 36, in <module>
  File "/usr/local/lib/python3.9/site-packages/netmiko/base_connection.py", line 1927, in send_config_set
    raise ConfigInvalidException(msg)
netmiko.ssh_exception.ConfigInvalidException: Invalid input detected at command: bgp cluster-id 0.0.17.2
Here the VRF did not exist, but it also applied the cluster-id value to the configuration. Could you point me on what I am doing wrong here? Ideally it shouldn't change the cluster id since the error should be detected after sending the second command, "address-family ipv4 vrf VRF_WL"

**Reply by ktbyers** [reply]
When you say `it also applied`, the `it` here is the Cisco device, not Netmiko:

```
R1(config-router)#exit
Assuming initial cluster id to be 0.0.16.0,      # This is Cisco doing this
```

You should be able to verify if you do replicate that behavior manually.

Let me know if you observe something different.

**Reply by kbSean1** [reply]
@ktbyers I executed it manually and saw that the bgp cluster-id got changed in this case(Which like you mentioned , Cisco is doing ) this one. 
But if we use  error_pattern feature in send_config_set , it should abort the next lines of configuration, if specific error messages are detected. Right? 
I am expecting to see this behavior, where , when I execute the second command, which is address-family ipv4 vrf VRF_WL, the device throws an error, for example, "% VRF VRF_WL does not exist." and then Netmiko should abort next line "bgp cluster-id 0.0.17.2" , from configuring right? Is there any way to solve this issue in Netmiko?

Is there  a way to do this using send_multiline(commands)?
